apps/rush-lib/src/logic/taskExecution/TaskExecutionManager.ts

```diff
@@ -174,7 +170,36 @@ export class TaskExecutionManager {
 
     this._terminal.writeStdoutLine(`Executing a maximum of ${this._parallelism} simultaneous processes...`);
 
-    await this._startAvailableTasksAsync();
+    const maxParallelism: number = Math.min(totalTasks, this._parallelism);
+    const taskPrioritySort: ITaskSortFunction = (a: Task, b: Task): number => {
+      let diff: number = a.criticalPathLength! - b.criticalPathLength!;
+      if (diff) {
+        return diff;
+      }
+
+      diff = a.dependents.size - b.dependents.size;
+      if (diff) {
+        return diff;
+      }
+
+      // No further default considerations.
+      return 0;
+    };
+    const taskQueue: AsyncTaskQueue = new AsyncTaskQueue(this._tasks, taskPrioritySort);
+
+    // Iterate in parallel with maxParallelism concurrent lanes
+    await Promise.all(
+      Array.from({ length: maxParallelism }, async (unused: undefined, index: number): Promise<void> => {
+        // laneId can be used in logging to examine concurrency
+        const laneId: number = index + 1;
+        // The taskQueue is a singular async iterable that stalls until a task is available, and marks itself
+        // done when the queue is empty.
+        for await (const task of taskQueue) {
+          // Take a task, execute it, wait for it to finish, wait for a new task
+          await this._executeTaskAsync(task, laneId);
+        }
+      })
+    );
```

Comment:
This feels fairly similar to `Async.forEachAsync(..., () => { ... }, { concurrency: maxParallelism });`

Reply:
It is similar, but this version was created to support the DevTools tracing style logging to keep track of which CPU core or "parallelism lane" a given task executes on; information which does not exist in the confines of `Async.forEachAsync`. By tracking this information it makes the parallelism at any point during the execution considerably easier to view. It also currently has slightly different error handling behavior, though that's more due to laziness than being a design objective.